nixos/modules/services/misc/matrix-synapse.nix

```diff
@@ -591,36 +629,57 @@ in {
 
   config = mkIf cfg.enable {
     users.extraUsers = [
-      { name = "matrix-synapse";
-        group = "matrix-synapse";
-        home = "/var/lib/matrix-synapse";
+      { name = user;
+        group = group;
+        home = "${cfg.state_path}";
         createHome = true;
         shell = "${pkgs.bash}/bin/bash";
-        uid = config.ids.uids.matrix-synapse;
+        uid = config.ids.uids."${user}";
       } ];
 
     users.extraGroups = [
-      { name = "matrix-synapse";
-        gid = config.ids.gids.matrix-synapse;
+      { name = group;
+        gid = config.ids.gids."${group}";
       } ];
 
+    services.postgresql.enable = mkIf usePostgresql (mkDefault true);
+
     systemd.services.matrix-synapse = {
       description = "Synapse Matrix homeserver";
-      after = [ "network.target" ];
+      after = [ "network.target" "postgresql.service" ];
       wantedBy = [ "multi-user.target" ];
       preStart = ''
         ${cfg.package}/bin/homeserver \
           --config-path ${configFile} \
-          --keys-directory /var/lib/matrix-synapse \
+          --keys-directory ${cfg.state_path} \
           --generate-keys
+      '' + optionalString (usePostgresql && cfg.create_local_database) ''
+        if ! test -e "${cfg.state_path}/db-created"; then
+          ${pkgs.sudo}/bin/sudo -u ${pg.superUser} \
+            ${pg.package}/bin/createuser \
+            --login \
+            --no-createdb \
+            --no-createrole \
+            --encrypted \
+            ${cfg.database_user}
+          ${pkgs.sudo}/bin/sudo -u ${pg.superUser} \
+            ${pg.package}/bin/createdb \
+            --owner=${cfg.database_user} \
+            --encoding=UTF8 \
+            --lc-collate=C \
+            --lc-ctype=C \
+            --template=template0 \
+            ${cfg.database_name}
```

Comment:
Are these db setup commands guaranteed to be idempotent?

Reply:
They will fail if you remove `db-created` without dropping the db/role.  It's not a very elegant solution, but it's what's used in other similar modules.

Reply:
Er, right, not sure how I managed to forget the guard file even though it's right there.